### ISSUE TITLE
[runtime][python] Should DeviceArray.to_host() use a cache?

### ISSUE THREAD

**Issue by sogartar** [iree-org/iree#18870]
`DeviceArray`'s `to_host()` method would reuse a [cached result](https://github.com/iree-org/iree/blob/b922a7012978b149e87fd828b893eb3a1a05485a/runtime/bindings/python/iree/runtime/array_interop.py#L125). In the mean time the contents of the array may have changed and subsequent calls to `to_host` will return the cached value, which at this point is wrong.
Is this the intended behavior? My expectation is that it should not cache anything and copy it every time if the array is not mappable to the host.

**Comment by stellaraccident** [comment]
Yes, that design is problematic. This was originally written to make some porting easier and includes too much implicit behavior. I would be fine removing those lines you reference. We could also add an option to force or something, but it is better to not add more weird behavior.
